_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,227 @@
+
+# Accessing Zookeeper
+
+Be default, communication between Zookeeper and all other Strimzi components must be encrytped. Using a hack by Jakub Scholz, one can create a Kubernetes service which reuses the Strimzi certs to pipe unecrypted Cruise Control traffic through to Zookeeper. This removes the need to include traffic encryting sidecars in the Cruise Control deployment.
+
+```
+kubectl apply -f https://gist.githubusercontent.com/scholzj/6cfcf9f63f73b54eaebf60738cfdbfae/raw/068d55ac65e27779f3a5279db96bae03cea70acb/zoo-entrance.yaml
+```
+
+# Containerizing and deploying Cruise Control
+
+Create a Cruise Control image
+
+```Dockerfile
+FROM centos:7
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
+
+RUN yum -y install git java-1.8.0-openjdk-devel && \
+    yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+WORKDIR cruise-control
+
+RUN ./gradlew jar copyDependantLibs
+
+RUN sed -i 's/bootstrap.servers=.*/bootstrap.servers=my-cluster-kafka-bootstrap:9092/g' config/cruisecontrol.properties \
+    && sed -i 's/zookeeper.connect=.*/zookeeper.connect=zoo-entrance:2181/g' config/cruisecontrol.properties \
+    && sed -i 's/capacity.config.file=.*/capacity.config.file=config\/capacityJBOD.json/g' config/cruisecontrol.properties \
+    && sed -i 's/sample.store.topic.replication.factor=.*/sample.store.topic.replication.factor=1/g' config/cruisecontrol.properties 
+   
+RUN mkdir logs \ 
+    && touch ./logs/kafkacruisecontrol.log \
+    && chmod a+rw -R .
+
+RUN curl -L https://github.com/linkedin/cruise-control-ui/releases/download/v0.1.0/cruise-control-ui.tar.gz \
+    -o /tmp/cruise-control-ui.tar.gz \
+    && tar zxvf /tmp/cruise-control-ui.tar.gz
+
+ENTRYPOINT ["/bin/bash", "-c", "./kafka-cruise-control-start.sh config/cruisecontrol.properties"]
+```
+
+```
+docker build . -t <cruise-control-image-name>
+```
+
+Then create a Cruise Control deployment, `cruise-control-deployment.yaml`
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: my-cruise-control
+  labels:
+    app: cruise-control
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: cruise-control
+        name: my-cruise-control
+    spec:
+      containers:
+      - name: my-cruise-control
+        image: <cruise-control-image-name>
+        imagePullPolicy: 'Always'
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app: cruise-control
+    name: my-cruise-control
+  name: my-cruise-control
+spec:
+  ports:
+    - name: http-9090
+      port: 9090
+      protocol: TCP
+      targetPort: 9090
+  type: NodePort
+  selector:
+    name: my-cruise-control
+
+```
+
+```bash
+oc apply -f cruise-control-deployment.yaml
+```
+
+# Interacting with Cruise Control API
+
+To access the Cruise Control REST API from outside out OKD, port-forward the Cruise Control service to localhost:9095:
+
+```
+oc port-forward svc/my-cruise-control 9095:9090
```

Comment:
Why not using 9090 even on localhost? Are you using in your test a port forwarding for Prometheus as well? I noticed that CC uses the same Prometheus port.

Reply:
I wanted to make a distinction between which port in the `oc` command was for the CC and which was for localhost. I wasn't using Prometheus so we could use 9090 for localhost here as well